plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	return (guint16)val;
 }
 
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device)
+{
+	GError **error_local_offline = NULL;
+	GError **error_local_rescan = NULL;
+
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/offline", "1", error_local_offline)) {
+		g_set_error_literal(error_local_offline,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "Setting usb4 port offline failed");
+		return FALSE;
+	}
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/rescan", "1", error_local_rescan)) {
+		g_set_error_literal(error_local_rescan,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "Rescan on port failed");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static void
+fu_thunderbolt_device_set_port_online(FuUdevDevice *device)
+{
+	FuUdevDevice *udev = FU_UDEV_DEVICE(device);
+	GError **error_local = NULL;
+
+	if (!fu_udev_device_write_sysfs(udev, "usb4_port1/offline", "0", error_local)) {
+		g_set_error_literal(error_local,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "Setting port online failed");
+	}
+}
+
+static FuUdevDevice *
+get_udev_parent_device(FuDevice *device)
+{
+	g_autoptr(GUdevDevice) udev_parent1 = NULL;
+	g_autoptr(GUdevDevice) udev_parent2 = NULL;
+	GUdevDevice *udev_device = NULL;
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER)
+		return NULL;
+	udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	if (udev_device == NULL)
+		return NULL;
+	udev_parent1 = g_udev_device_get_parent(udev_device);
+	if (udev_parent1 == NULL)
+		return NULL;
+	udev_parent2 = g_udev_device_get_parent(udev_parent1);
+	return fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
+					       g_steal_pointer(&udev_parent2));
+}
+gboolean
+fu_thunderbolt_device_open(FuDevice *device, GError **error)
```

Comment:
I'd suggest to rename it to `fu_thunderbolt_retimer_open` as this is only used in the retimer case